runtime/src/bank.rs

```diff
@@ -8115,30 +8135,81 @@ impl Bank {
-                self.capitalization
-                    .fetch_sub(old_account.lamports(), Relaxed);
-
-                // Transfer new account to old account
-                self.store_account(old_address, &new_account);
+                // Derive each program's data account address (PDA)
+                let (old_data_address, _) = Pubkey::find_program_address(
+                    &[old_address.as_ref()],
+                    &bpf_loader_upgradeable::id(),
+                );
+                let (new_data_address, _) = Pubkey::find_program_address(
+                    &[new_address.as_ref()],
+                    &bpf_loader_upgradeable::id(),
+                );
 
-                // Clear new account
-                self.store_account(new_address, &AccountSharedData::default());
+                // If a data account is also provided with this new program
+                // account, then we want to update the existing data account
+                if let Some(new_data_account) = self.get_account_with_fixed_root(&new_data_address)
+                {
+                    // A data account exists for the new program
+                    // Check if the old program account has a data account
+                    if let Some(old_data_account) =
+                        self.get_account_with_fixed_root(&old_data_address)
+                    {
+                        // It does. Replace it with the new data account
+                        self.replace_account(
+                            &old_data_address,
+                            &new_data_address,
+                            Some(&old_data_account),
+                            &new_data_account,
+                        );
+                        // The old program account will already house the PDA
+                        // of the data account
+                    } else {
+                        // It does _not_. Create it with the new data account
+                        self.replace_account(
+                            &old_data_address,
+                            &new_data_address,
+                            None,
+                            &new_data_account,
+                        );
+                        // Update the old program account to house the PDA of
+                        // the data account
+                        {
+                            let mut account = Account::from(old_account.clone());
+                            account.data = old_data_address.as_ref().to_vec();
+                            let account_shared_data = AccountSharedData::from(account);
+                            self.store_account(old_address, &account_shared_data);
+                        }
+                    }
+                    // We only swapped the data accounts, so now we need to
+                    // clear the new program account
+                    self.capitalization
+                        .fetch_sub(new_account.lamports(), Relaxed);
```

Comment:
Heh, *maybe*

Reply:
Only the proposed new data account was used, so do we leave this account alone or clear it? After all, it should just be a pointer to the now-deceased data account, which is now in place of the old program data account

Reply:
Ah sorry, this was in reference to my previous comment -- if we end up stealing the lamports from the previous account, we don't want to subtract those from the total capitalization.